File: backend/src/config/env.ts
// Environment variable validation and configuration
// Uses Zod for runtime type checking

import { config } from 'dotenv';
import { z } from 'zod';

// Load environment variables from .env file
config();

// Define the schema for environment variables
const envSchema = z.object({
  // Database
  DATABASE_URL: z.string().url(),

  // JWT
  JWT_SECRET: z.string().min(32, 'JWT_SECRET must be at least 32 characters'),
  JWT_EXPIRES_IN: z.string().default('7d'),

  // Server
  PORT: z.string().transform(Number).default('3000'),
  NODE_ENV: z.enum(['development', 'production', 'test']).default('development'),

  // CORS
  CORS_ORIGIN: z.string().default('http://localhost:5173'),

  // Rate Limiting
  RATE_LIMIT_WINDOW_MS: z.string().transform(Number).default('900000'),
  RATE_LIMIT_MAX_REQUESTS: z.string().transform(Number).default('100'),

  // Stripe
  STRIPE_SECRET_KEY: z.string().optional(),
  STRIPE_PUBLISHABLE_KEY: z.string().optional(),
});

// Parse and validate environment variables
const parseEnv = () => {
  try {
    return envSchema.parse(process.env);
  } catch (error) {
    if (error instanceof z.ZodError) {
      console.error('❌ Environment variable validation failed:');
      error.errors.forEach((err) => {
        console.error(`  - ${err.path.join('.')}: ${err.message}`);
      });
      process.exit(1);
    }
    throw error;
  }
};

export const env = parseEnv();

// Export typed environment variables
export const {
  DATABASE_URL,
  JWT_SECRET,
  JWT_EXPIRES_IN,
  PORT,
  NODE_ENV,
  CORS_ORIGIN,
  RATE_LIMIT_WINDOW_MS,
  RATE_LIMIT_MAX_REQUESTS,
  STRIPE_SECRET_KEY,
  STRIPE_PUBLISHABLE_KEY,
} = env;

// Helper functions
export const isDevelopment = () => NODE_ENV === 'development';
export const isProduction = () => NODE_ENV === 'production';
export const isTest = () => NODE_ENV === 'test';

File: backend-go/internal/config/env.go
package config

import (
	"log"
	"os"
	"strconv"

	"github.com/joho/godotenv"
)

// Config holds all application configuration
type Config struct {
	Port                  string
	Environment           string
	DatabaseURL           string
	JWTSecret             string
	StripeSecretKey       string
	StripePublishableKey  string
	CORSOrigin            string
	RateLimitWindowMS     int
	RateLimitMaxRequests  int
}

var AppConfig *Config

// LoadConfig loads environment variables and initializes application config
func LoadConfig() *Config {
	// Load .env file (ignore error in production)
	godotenv.Load()

	config := &Config{
		Port:                  getEnv("PORT", "3001"),
		Environment:           getEnv("GO_ENV", "development"),
		DatabaseURL:           getEnv("DATABASE_URL", "postgresql://postgres:postgres@localhost:5433/ecommerce?sslmode=disable"),
		JWTSecret:             getEnv("JWT_SECRET", "default-secret-change-in-production"),
		StripeSecretKey:       getEnv("STRIPE_SECRET_KEY", ""),
		StripePublishableKey:  getEnv("STRIPE_PUBLISHABLE_KEY", ""),
		CORSOrigin:            getEnv("CORS_ORIGIN", "http://localhost:5173"),
		RateLimitWindowMS:     getEnvAsInt("RATE_LIMIT_WINDOW_MS", 900000),
		RateLimitMaxRequests:  getEnvAsInt("RATE_LIMIT_MAX_REQUESTS", 100),
	}

	// Validate required fields
	if config.JWTSecret == "default-secret-change-in-production" {
		log.Println("⚠️  WARNING: Using default JWT secret. Set JWT_SECRET in production!")
	}

	AppConfig = config
	return config
}

// IsDevelopment returns true if running in development mode
func (c *Config) IsDevelopment() bool {
	return c.Environment == "development"
}

// IsProduction returns true if running in production mode
func (c *Config) IsProduction() bool {
	return c.Environment == "production"
}

// Helper functions
func getEnv(key, defaultValue string) string {
	if value := os.Getenv(key); value != "" {
		return value
	}
	return defaultValue
}

func getEnvAsInt(key string, defaultValue int) int {
	valueStr := os.Getenv(key)
	if value, err := strconv.Atoi(valueStr); err == nil {
		return value
	}
	return defaultValue
}
